nifi-nar-bundles/nifi-framework-bundle/nifi-stateless-bundle/README.md

```diff
@@ -0,0 +1,405 @@
+|---------|------------------|----------------|
+| Data Durability | Data is reliably stored on disk in the FlowFile and Content Repositories | Data is stored in-memory and must be consumed from the source again upon restart |
+| Data Ordering | Data is ordered independently in each Connection based on the selected Prioritizers | Data flows through the system in the order it was received (First-In, First-Out / FIFO) |
+| Site-to-Site | Supports full Site-to-Site capabilities, including Server and Client roles | Can push to, or pull from, a NiFi instance but cannot receive incoming Site-to-Site connections. I.e., works as a client but not a server. |
+| Form Factor | Large form factor. Designed to take advantage of many cores and disks. | Light-weight form factor. Easily embedded into another application. Single-threaded processing. |
+| Heap Considerations | Typically, many processors in use by many users. FlowFile content should not be loaded into heap because it can easily cause heap exhaustion. | Smaller dataflows use less heap. Flow operates on only one or a few FlowFiles at a time and holds FlowFile contents in memory in the Java heap. |
+| Data Provenance | Fully stored, indexed data provenance that can be browsed through the UI and exported via Reporting Tasks | Limited Data Provenance capabilities, events being stored in memory. No ability to view but can be exported using Reporting Tasks. However, since they are in-memory, they will be lost upon restart and may roll off before they can be exported. |
+| Embeddability | While technically possible to embed traditional NiFi, it is not recommended, as it launches a heavy-weight User Interface, deals with complex authentication and authorization, and several file-based external dependencies, which can be difficult to manage. | Has minimal external dependencies (directory containing extensions and a working directory to use for temporary storage) and is much simpler to manage. Embeddability is an important feature of Stateless NiFi. |
+ 
+## Running Stateless NiFi
+
+Stateless NiFi can be used as a library and embedded into other applications. However, it can also be run directly
+from the command-line from a NiFi build using the `bin/nifi.sh` script.
+
+To do so requires three files:
+
+- The engine configuration properties file
+- The dataflow configuration properties file
+- The dataflow itself (which may exist as a file, or point to a flow in a NiFi registry)
+
+Stateless NiFi accepts two separate configuration files: an engine configuration file and a dataflow configuration file.
+This is done because typically the engine configuration will be the same for all flows that are run, so it can be created
+only once. The dataflow configuration will be different for each dataflow that is to be run.
+
+An example of running stateless NiFi:
+
+```
+bin/nifi.sh stateless -c /var/lib/nifi/stateless/config/stateless.properties /var/lib/nifi/stateless/flows/jms-to-kafka.properties
```

Comment:
Is there a particular reason for using positional arguments for the engine and data flow properties?  For the use case where the data flow is provided as a JSON file, it would be helpful to support passing that file as a direct argument as opposed to requiring the specification of the data flow properties file.  Either way, if the engine and data flow properties were specified using arguments, it seems like it would be easier to support other optional arguments in future versions.

Reply:
They are required arguments, not options. While it's common to see, I very much dislike the notion of using command line flags for required arguments. I can see some argument for the idea of specifying a dataflow JSON file as a single argument instead of a dataflow properties file, and I did consider that option. But that would only be applicable for cases where the user wants to specify all required parameters from command-line and have no reporting tasks. It would also make the command less portable if other required properties were to be added for the dataflow properties. All in all, I think it would lead to more confusion than convenience.

Reply:
Thanks for the explanation, that makes sense.  As implemented, it does provide a helpful differentiation between required arguments and options, so that is helpful.